<center>
<img src="https://habrastorage.org/web/677/8e1/337/6778e1337c3d4b159d7e99df94227cb2.jpg"/>
## Специализация "Машинное обучение и анализ данных"
</center>
<center>Автор материала: программист-исследователь Mail.ru Group, старший преподаватель Факультета Компьютерных Наук ВШЭ Юрий Кашницкий

# <center> Capstone проект №1. Идентификация пользователей по посещенным веб-страницам

В этом проекте мы будем решать задачу идентификации пользователя по его поведению в сети Интернет. Это сложная и интересная задача на стыке анализа данных и поведенческой психологии. В качестве примера, компания Яндекс решает задачу идентификации взломщика почтового ящика по его поведению. В двух словах, взломщик будет себя вести не так, как владелец ящика: он может не удалять сообщения сразу по прочтении, как это делал хозяин, он будет по-другому ставить флажки сообщениям и даже по-своему двигать мышкой. Тогда такого злоумышленника можно идентифицировать и "выкинуть" из почтового ящика, предложив хозяину войти по SMS-коду. Этот пилотный проект описан в [статье](https://habrahabr.ru/company/yandex/blog/230583/) на Хабрахабре. Похожие вещи делаются, например, в Google Analytics и описываются в научных статьях, найти можно многое по фразам "Traversal Pattern Mining" и "Sequential Pattern Mining".

<img src='http://i.istockimg.com/file_thumbview_approve/21546327/5/stock-illustration-21546327-identification-de-l-utilisateur.jpg'>

Мы будем решать похожую задачу: по последовательности из нескольких веб-сайтов, посещенных подряд один и тем же человеком, мы будем идентифицировать этого человека. Идея такая: пользователи Интернета по-разному переходят по ссылкам, и это может помогать их идентифицировать (кто-то сначала в почту, потом про футбол почитать, затем новости, контакт, потом наконец – работать, кто-то – сразу работать).

Будем использовать данные из [статьи](http://ceur-ws.org/Vol-1703/paper12.pdf) "A Tool for Classification of Sequential Data". И хотя мы не можем рекомендовать эту статью (описанные методы делеки от state-of-the-art, лучше обращаться к [книге](http://www.charuaggarwal.net/freqbook.pdf) "Frequent Pattern Mining" и последним статьям с ICDM), но данные там собраны аккуратно и представляют интерес.

Имеются данные с прокси-серверов Университета Блеза Паскаля, они имеют очень простой вид. Для каждого пользователя заведен csv-файл с названием user\*\*\*\*.csv (где вместо звездочек – 4 цифры, соответствующие ID пользователя), а в нем посещения сайтов записаны в следующем формате: <br>

<center>*timestamp, посещенный веб-сайт*</center>

Скачать исходные данные можно по ссылке в статье, там же описание.
Для этого задания хватит данных не по всем 3000 пользователям, а по 10 и 150. [Ссылка](https://yadi.sk/d/3gscKIdN3BCASG) на архив *capstone_user_identification* (~7 Mb, в развернутом виде ~ 60 Mb). 

В финальном проекте уже придется столкнуться с тем, что не все операции можно выполнить за разумное время (скажем, перебрать с кросс-валидацией 100 комбинаций параметров случайного леса на этих данных Вы вряд ли сможете), поэтому мы будем использовать параллельно 2 выборки: по 10 пользователям и по 150. Для 10 пользователей будем писать и отлаживать код, для 150 – будет рабочая версия. 

Данные устроены следующем образом:

 - В каталоге 10users лежат 10 csv-файлов с названием вида "user[USER_ID].csv", где [USER_ID] – ID пользователя;
 - Аналогично для каталога 150users – там 150 файлов;
 - В каталоге 3users – игрушечный пример из 3 файлов, это для отладки кода предобработки, который Вы далее напишете.

На 5 неделе будет задание по [соревнованию](https://inclass.kaggle.com/c/identify-me-if-you-can4) Kaggle Inclass, которое организовано специально под Capstone проект нашей специализации. Соревнование уже открыто и, конечно, желающие могут начать уже сейчас.

# <center>Неделя 1. Подготовка данных к анализу и построению моделей

Первая часть проекта посвящена подготовке данных для дальнейшего описательного анализа и построения прогнозных моделей. Надо будет написать код для предобработки данных (исходно посещенные веб-сайты указаны для каждого пользователя в отдельном файле) и формирования единой обучающей выборки. Также в этой части мы познакомимся с разреженным форматом данных (матрицы `Scipy.sparse`), который хорошо подходит для данной задачи. 

**План 1 недели:**
 - Часть 1. Подготовка обучающей выборки
 - Часть 2. Работа с разреженным форматом данных

<font color='red'>**Задание:**</font> заполните код в этой тетрадке и выберите ответы в [веб-форме](https://docs.google.com/forms/d/e/1FAIpQLSedmwHb4cOI32zKJmEP7RvgEjNoz5GbeYRc83qFXVH82KFgGA/viewform). 


**В этой части проекта Вам могут быть полезны видеозаписи следующих лекций 1 и 2 недели курса "Математика и Python для анализа данных":**
   - [Циклы, функции, генераторы, list comprehension](https://www.coursera.org/learn/mathematics-and-python/lecture/Kd7dL/tsikly-funktsii-ghienieratory-list-comprehension)
   - [Чтение данных из файлов](https://www.coursera.org/learn/mathematics-and-python/lecture/8Xvwp/chtieniie-dannykh-iz-failov)
   - [Запись файлов, изменение файлов](https://www.coursera.org/learn/mathematics-and-python/lecture/vde7k/zapis-failov-izmienieniie-failov)
   - [Pandas.DataFrame](https://www.coursera.org/learn/mathematics-and-python/lecture/rcjAW/pandas-data-frame)
   - [Pandas. Индексация и селекция](https://www.coursera.org/learn/mathematics-and-python/lecture/lsXAR/pandas-indieksatsiia-i-sieliektsiia)
   
**Кроме того, в задании будут использоваться библиотеки Python [`glob`](https://docs.python.org/3/library/glob.html), [`pickle`](https://docs.python.org/2/library/pickle.html) и класс [`csr_matrix`](https://docs.scipy.org/doc/scipy-0.18.1/reference/generated/scipy.sparse.csr_matrix.html) из `Scipy.sparse`.**

Наконец, для лучшей воспроизводимости результатов приведем список версий основных используемых в проекте библиотек: NumPy, SciPy, Pandas, Matplotlib, Statsmodels и Scikit-learn. Для этого воспользуемся расширением [watermark](https://github.com/rasbt/watermark).

In [2]:
# pip install watermark
%load_ext watermark

In [2]:
#read-only
%watermark -v -m -p numpy,scipy,pandas,matplotlib,statsmodels,sklearn -g

CPython 3.6.1
IPython 5.3.0

numpy 1.13.1
scipy 0.19.1
pandas 0.20.3
matplotlib 2.0.0
statsmodels 0.8.0rc1
sklearn 0.18.2

compiler   : GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)
system     : Darwin
release    : 16.7.0
machine    : x86_64
processor  : i386
CPU cores  : 8
interpreter: 64bit
Git hash   : 1f49f55605dfc4979e22e6ba6a13d06b8921b8b6


In [3]:
%watermark -v -m -p numpy,scipy,pandas,matplotlib,statsmodels,sklearn -g

CPython 3.6.3
IPython 6.1.0

numpy 1.13.3
scipy 0.19.1
pandas 0.20.3
matplotlib 2.1.0
statsmodels 0.8.0
sklearn 0.19.1

compiler   : MSC v.1900 64 bit (AMD64)
system     : Windows
release    : 10
machine    : AMD64
processor  : Intel64 Family 6 Model 42 Stepping 7, GenuineIntel
CPU cores  : 8
interpreter: 64bit
Git hash   : f153536d4a2bbc9582b8f5e96fcd054b84713427


In [5]:
from __future__ import division, print_function
# отключим всякие предупреждения Anaconda
import warnings
warnings.filterwarnings('ignore')
from glob import glob
import os
import pickle
#pip install tqdm
from tqdm import tqdm_notebook
import numpy as np
import pandas as pd
from scipy.sparse import csr_matrix

**Посмотрим на один из файлов с данными о посещенных пользователем (номер 31) веб-страницах.**

In [6]:
# Поменяйте на свой путь к данным
PATH_TO_DATA = '../../data/capstone_user_identification'

In [3]:
!ls $PATH_TO_DATA

10users
150users
3users
__MACOSX


In [7]:
user31_data = pd.read_csv(os.path.join(PATH_TO_DATA, 
                                       '10users/user0031.csv'))

In [8]:
user31_data.head()

timestamp                        site
0  2013-11-15 08:12:07  fpdownload2.macromedia.com
1  2013-11-15 08:12:17                 laposte.net
2  2013-11-15 08:12:17             www.laposte.net
3  2013-11-15 08:12:17              www.google.com
4  2013-11-15 08:12:18             www.laposte.net

**Поставим задачу классификации: идентифицировать пользователя по сессии из 10 подряд посещенных сайтов. Объектом в этой задаче будет сессия из 10 сайтов, последовательно посещенных одним и тем же пользователем, признаками – индексы этих 10 сайтов (чуть позже здесь появится "мешок" сайтов, подход Bag of Words). Целевым классом будет id пользователя.**

### <center>Пример для иллюстрации</center>
**Пусть пользователя всего 2, длина сессии – 2 сайта.**

<center>user0001.csv</center>
<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;}
.tg .tg-yw4l{vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-031e">timestamp</th>
    <th class="tg-031e">site</th>
  </tr>
  <tr>
    <td class="tg-031e">00:00:01</td>
    <td class="tg-031e">vk.com</td>
  </tr>
  <tr>
    <td class="tg-yw4l">00:00:11</td>
    <td class="tg-yw4l">google.com</td>
  </tr>
  <tr>
    <td class="tg-031e">00:00:16</td>
    <td class="tg-031e">vk.com</td>
  </tr>
  <tr>
    <td class="tg-031e">00:00:20</td>
    <td class="tg-031e">yandex.ru</td>
  </tr>
</table>

<center>user0002.csv</center>
<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;}
.tg .tg-yw4l{vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-031e">timestamp</th>
    <th class="tg-031e">site</th>
  </tr>
  <tr>
    <td class="tg-031e">00:00:02</td>
    <td class="tg-031e">yandex.ru</td>
  </tr>
  <tr>
    <td class="tg-yw4l">00:00:14</td>
    <td class="tg-yw4l">google.com</td>
  </tr>
  <tr>
    <td class="tg-031e">00:00:17</td>
    <td class="tg-031e">facebook.com</td>
  </tr>
  <tr>
    <td class="tg-031e">00:00:25</td>
    <td class="tg-031e">yandex.ru</td>
  </tr>
</table>

Идем по 1 файлу, нумеруем сайты подряд: vk.com – 1, google.com – 2 и т.д. Далее по второму файлу. 

Отображение сайтов в их индесы должно получиться таким:

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;}
.tg .tg-yw4l{vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-031e">site</th>
    <th class="tg-yw4l">site_id</th>
  </tr>
  <tr>
    <td class="tg-yw4l">vk.com</td>
    <td class="tg-yw4l">1</td>
  </tr>
  <tr>
    <td class="tg-yw4l">google.com</td>
    <td class="tg-yw4l">2</td>
  </tr>
  <tr>
    <td class="tg-yw4l">yandex.ru</td>
    <td class="tg-yw4l">3</td>
  </tr>
  <tr>
    <td class="tg-yw4l">facebook.com</td>
    <td class="tg-yw4l">4</td>
  </tr>
</table>

Тогда обучающая выборка будет такой (целевой признак – user_id):
<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;}
.tg .tg-s6z2{text-align:center}
.tg .tg-baqh{text-align:center;vertical-align:top}
.tg .tg-hgcj{font-weight:bold;text-align:center}
.tg .tg-amwm{font-weight:bold;text-align:center;vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-hgcj">session_id</th>
    <th class="tg-hgcj">site1</th>
    <th class="tg-hgcj">site2</th>
    <th class="tg-amwm">user_id</th>
  </tr>
  <tr>
    <td class="tg-s6z2">1</td>
    <td class="tg-s6z2">1</td>
    <td class="tg-s6z2">2</td>
    <td class="tg-baqh">1</td>
  </tr>
  <tr>
    <td class="tg-s6z2">2</td>
    <td class="tg-s6z2">1</td>
    <td class="tg-s6z2">3</td>
    <td class="tg-baqh">1</td>
  </tr>
  <tr>
    <td class="tg-s6z2">3</td>
    <td class="tg-s6z2">3</td>
    <td class="tg-s6z2">2</td>
    <td class="tg-baqh">2</td>
  </tr>
  <tr>
    <td class="tg-s6z2">4</td>
    <td class="tg-s6z2">4</td>
    <td class="tg-s6z2">3</td>
    <td class="tg-baqh">2</td>
  </tr>
</table>

Здесь 1 объект – это сессия из 2 посещенных сайтов 1-ым пользователем (target=1). Это сайты vk.com и google.com (номер 1 и 2). И так далее, всего 4 сессии. Пока сессии у нас не пересекаются по сайтам, то есть посещение каждого отдельного сайта относится только к одной сессии.

## Часть 1. Подготовка обучающей выборки
Реализуйте функцию *prepare_train_set*, которая принимает на вход путь к каталогу с csv-файлами *path_to_csv_files* и параметр *session_length* – длину сессии, а возвращает 2 объекта:
- DataFrame, в котором строки соответствуют уникальным сессиям из *session_length* сайтов, *session_length* столбцов – индексам этих *session_length* сайтов и последний столбец – ID пользователя
- частотный словарь сайтов вида {'site_string': [site_id, site_freq]}, например для недавнего игрушечного примера это будет {'vk.com': (1, 2), 'google.com': (2, 2), 'yandex.ru': (3, 3), 'facebook.com': (4, 1)}

Детали:
- Смотрите чуть ниже пример вывода, что должна возвращать функция
- Используйте `glob` (или аналоги) для обхода файлов в каталоге. Для определенности, отсортируйте список файлов лексикографически. Удобно использовать `tqdm_notebook` (или просто `tqdm` в случае python-скрипта) для отслеживания числа выполненных итераций цикла
- Создайте частотный словарь уникальных сайтов (вида {'site_string': (site_id, site_freq)}) и заполняйте его по ходу чтения файлов. Начните с 1
- Рекомендуется меньшие индексы давать более часто попадающимся сайтам (приницип наименьшего описания)
- Не делайте entity recognition, считайте *google.com*, *http://www.google.com* и *www.google.com* разными сайтами (подключить entity recognition можно уже в рамках индивидуальной работы над проектом)
- Скорее всего в файле число записей не кратно числу *session_length*. Тогда последняя сессия будет короче. Остаток заполняйте нулями. То есть если в файле 24 записи и сессии длины 10, то 3 сессия будет состоять из 4 сайтов, и ей мы сопоставим вектор [*site1_id*, *site2_id*, *site3_id*, *site4_id*, 0, 0, 0, 0, 0, 0, *user_id*] 
- В итоге некоторые сессии могут повторяться – оставьте как есть, не удаляйте дубликаты. Если в двух сессиях все сайты одинаковы, но сессии принадлежат разным пользователям, то тоже оставляйте как есть, это естественная неопределенность в данных
- Не оставляйте в частотном словаре сайт 0 (уже в конце, когда функция возвращает этот словарь)
- 150 файлов из *capstone_websites_data/150users/* у меня обработались за 1.7 секунды, но многое, конечно, зависит от реализации функции и от используемого железа. И вообще, первая реализация скорее всего будет не самой эффективной, дальше можно заняться профилированием (особенно если планируете запускать этот код для 3000 пользователей). Также эффективная реализация этой функции поможет нам на следующей неделе.

In [9]:
!ls $PATH_TO_DATA

10users
150users
3users
__MACOSX


In [12]:
path_to_csv_files = PATH_TO_DATA+'/3users'

In [13]:
!ls $path_to_csv_files

user0001.csv
user0002.csv
user0003.csv


In [16]:
!ls {os.path.join(PATH_TO_DATA, '3users')}

user0001.csv
user0002.csv
user0003.csv


In [95]:
def prepare_train_set(path_to_csv_files, session_length=10):
    ''' ВАШ КОД ЗДЕСЬ '''
    dict_unique_sites = {}
    columns = list()
    for i in range(session_length):
        columns.append('site'+str(i+1))
    columns.append('user_id')
    df = pd.DataFrame(columns=columns)
    site_index = 1
    for filename in glob(path_to_csv_files+'/*.csv'):
#         print(filename)
        user_id = int(filename[-8:-4])
        data = pd.read_csv(filename)
        session_number = 0   #положение сайта в session строке
        row = [0]*10         #session строка
        row.append(user_id)
        for site in data.site:
#             print (site)
            if site not in dict_unique_sites:
                dict_unique_sites[site]=(site_index, 1)
                site_index+=1
            else:
                dict_unique_sites[site] = dict_unique_sites[site][0],dict_unique_sites[site][1]+1
            if (session_number<session_length):
                row[session_number] = dict_unique_sites[site][0]
                session_number+=1
            else:
                df = df.append(pd.DataFrame([row], columns=columns), ignore_index=True)
                row = [0]*10
                row.append(user_id)
                session_number=0
                row[session_number] = dict_unique_sites[site][0]
                session_number+=1
        df = df.append(pd.DataFrame([row], columns=columns), ignore_index=True)
    return df, dict_unique_sites

In [4]:
train_data_toy, site_freq_3users = prepare_train_set(os.path.join(PATH_TO_DATA, '3users'), 
                                                     session_length=10)

In [138]:
train_data_toy

site1 site2 site3 site4 site5 site6 site7 site8 site9 site10 user_id
0     1     2     2     3     2     4     5     6     7      8       1
1     1     4     4     4     0     0     0     0     0      0       1
2     1     2     9     9     2     0     0     0     0      0       2
3    10     4     2     4     2     4     4     6    11     10       3
4    10     4     2     0     0     0     0     0     0      0       3

In [95]:
columns = list()
for i in range(1,10):
    columns.append('site'+str(i))
df = pd.DataFrame(columns=columns)

In [98]:
df

site1 site2 site3 site4 site5 site6 site7 site8 site9
0     0     1     2     3     4     5     6     7     8
1     0     1     2     3     4     5     6     7     8

In [97]:
row = [0,1,2,3,4,5,6,7,8]
df = df.append(pd.DataFrame([row], columns=columns), ignore_index=True)
df = df.append(pd.DataFrame([[0,1,2,3,4,5,6,7,8]], columns=columns), ignore_index=True)

In [139]:
prepare_train_set(os.path.join(PATH_TO_DATA, '3users'), 
                                                     session_length=10)

(  site1 site2 site3 site4 site5 site6 site7 site8 site9 site10 user_id
 0     1     2     2     3     2     4     5     6     7      8       1
 1     1     4     4     4     0     0     0     0     0      0       1
 2     1     2     9     9     2     0     0     0     0      0       2
 3    10     4     2     4     2     4     4     6    11     10       3
 4    10     4     2     0     0     0     0     0     0      0       3,
 {'accounts.google.com': (5, 1),
  'apis.google.com': (7, 1),
  'football.kulichki.ru': (9, 2),
  'geo.mozilla.org': (3, 1),
  'google.com': (4, 9),
  'mail.google.com': (6, 2),
  'meduza.io': (10, 3),
  'oracle.com': (2, 8),
  'plus.google.com': (8, 1),
  'vk.com': (1, 3),
  'yandex.ru': (11, 1)})

**Примените полученную функцию к игрушечному примеру, убедитесь, что все работает как надо.**

In [4]:
!cat $PATH_TO_DATA/3users/user0001.csv

timestamp,site
2013-11-15 09:28:17,vk.com
2013-11-15 09:33:04,oracle.com
2013-11-15 09:52:48,oracle.com
2013-11-15 11:37:26,geo.mozilla.org
2013-11-15 11:40:32,oracle.com
2013-11-15 11:40:34,google.com
2013-11-15 11:40:35,accounts.google.com
2013-11-15 11:40:37,mail.google.com
2013-11-15 11:40:40,apis.google.com
2013-11-15 11:41:35,plus.google.com
2013-11-15 12:40:35,vk.com
2013-11-15 12:40:37,google.com
2013-11-15 12:40:40,google.com
2013-11-15 12:41:35,google.com


In [5]:
!cat $PATH_TO_DATA/3users/user0002.csv

timestamp,site
2013-11-15 09:28:17,vk.com
2013-11-15 09:33:04,oracle.com
2013-11-15 09:52:48,football.kulichki.ru
2013-11-15 11:37:26,football.kulichki.ru
2013-11-15 11:40:32,oracle.com


In [10]:
!cat $PATH_TO_DATA/3users/user0003.csv

timestamp,site
2013-11-15 09:28:17,meduza.io
2013-11-15 09:33:04,google.com
2013-11-15 09:52:48,oracle.com
2013-11-15 11:37:26,google.com
2013-11-15 11:40:32,oracle.com
2013-11-15 11:40:34,google.com
2013-11-15 11:40:35,google.com
2013-11-15 11:40:37,mail.google.com
2013-11-15 11:40:40,yandex.ru
2013-11-15 11:41:35,meduza.io
2013-11-15 12:28:17,meduza.io
2013-11-15 12:33:04,google.com
2013-11-15 12:52:48,oracle.com


In [149]:
%%time
train_data_toy, site_freq_3users = prepare_train_set(os.path.join(PATH_TO_DATA, '3users'), 
                                                     session_length=10)

Wall time: 10 ms


In [146]:
%%timeit
train_data_toy, site_freq_3users = prepare_train_set(os.path.join(PATH_TO_DATA, '3users'), 
                                                     session_length=10)

8.85 ms ± 60.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [17]:
%%time
train_data_toy, site_freq_10users = prepare_train_set(os.path.join(PATH_TO_DATA, '10users'), 
                                                     session_length=10)

Wall time: 23.1 s


In [18]:
train_data_toy

site1 site2 site3 site4 site5 site6 site7 site8 site9 site10 user_id
0         1     2     3     4     3     3     4     3     5      3      31
1         6     7     8     9     3    10    11    12    13     14      31
2        14     4    14    14    15    16     6    17    18     14      31
3        19    20    19    14    14    14    14    21    22     23      31
4        24    14    15    25    26    27    28    29    30     29      31
5        31    30    27    27    14    14    32    33    34     35      31
6        36     2     3     4     3    10     3    11     1      3      31
7        37    38    39    11    14     4    14     4    22     23      31
8        14    17    40     5    41    42    14    43    40     15      31
9        44    36    45    36    46    47    48    48    47     49      31
10       49    33    33    50    51    51    52     4    51     53      31
11       53    54    55    56    57    57    58    59    51     59      31
12       60    61    62    63    64    60    65    66    59     67      31
13       59    68    15    69    70    71    71    59    61     10      31
14       72    10    72    59    15    58    52     4    51     51      31
15       73    51    74    75    73    75    74    73    74     76      31
16       73    74    73    76    77    77    77    77    78     79      31
17       79    78    79    80    78    81    81    80    79     78      31
18       82    82    79    78    83    79    84    85    84     79      31
19       83    85    84    84    79    78    79    78    85     79      31
20       78    78    81     4    79    78    51    81    86     87      31
21       78    79     4    88    79    51    86    89    88     78      31
22       79    88    79    89    80    78    88    79    89     90      31
23       91    78    79    86    88    88    79    86    89     86      31
24       89    79    89    92    93    89    79    89    88     86      31
25       89    78    79    86    89    86    79    86    89     89      31
26       88    86    86    86     7    86     8    86    24     32      31
27       86    86    86    89    86    16    89    94    95     50      31
28       51    51     4    52    51    96    97    96    98     99      31
29       32     5    94   100    15    98    96    42   101     98      31
...     ...   ...   ...   ...   ...   ...   ...   ...   ...    ...     ...
14031   567   587   567   577   575   576   576   578   566   1494     241
14032  1494  1494   575   576   576     9   576   575   575    576     241
14033   575   576   576   575   580   569   568   579   568    628     241
14034   629   629   567   566   758   630   576   575   629    575     241
14035   576   576   575   576   575   576   568    51    50    569     241
14036   568    51  2036  2036  2036    51   433   433    10      4     241
14037   433    10    51  2342     4   433   434     4  2342    433     241
14038   433     4    15   568   433    51    63    66   532   2342     241
14039    70    64    15   568     4   532   594   534   532      4     241
14040   532   594   319   568   532   533   532   211   532    532     241
14041   604    15   100  1347    51  1349  1350  1348  1868     10     241
14042   132  1888  1869  1868   211   565  1869   566  1870   1869     241
14043    33  1872    72   823     4    51    52   184   535   4913     241
14044  2029  2029    64    66    15  2029    63  2029   283    283     241
14045  1947   205   578  1871  1493  1493   576   575   575    576     241
14046   585   585   586   576   575   569   568    51    51    568     241
14047   568   568   568   568   579   580   628   629   567    758     241
14048   630   758    51  3585   185   186    52   184     4    188     241
14049   187   189   575   576   615   187    52   191   183    192     241
14050   187    51   575   576    51    51  1300  1300  1300   1300     241
14051  1300  1300  1300  1300  1300  1300  1300  1300  1300   1300     241
14052  1300  1300    51   187  

In [150]:
%%time
train_data_toy, site_freq_3users = prepare_train_set(os.path.join(PATH_TO_DATA, '150users'), 
                                                     session_length=10)

Wall time: 30min 35s


In [152]:
train_data_toy.to_csv('150users_train_data_v1_30m35s.csv', sep='\t')

In [154]:
df = pd.DataFrame.from_dict(site_freq_3users, orient="index")

In [156]:
df.to_csv('150users_site_freq_3users_v1_30m35s.csv', sep='\t')

Частоты сайтов (второй элемент кортежа) точно должны быть такими, нумерация может быть любой (первые элементы кортежей могут отличаться).

In [120]:
site_freq_3users

{'accounts.google.com': (5, 1),
 'apis.google.com': (7, 1),
 'football.kulichki.ru': (9, 2),
 'geo.mozilla.org': (3, 1),
 'google.com': (4, 9),
 'mail.google.com': (6, 2),
 'meduza.io': (10, 3),
 'oracle.com': (2, 8),
 'plus.google.com': (8, 1),
 'vk.com': (1, 3),
 'yandex.ru': (11, 1)}

In [14]:
site_freq_3users

{'accounts.google.com': (7, 1),
 'apis.google.com': (8, 1),
 'football.kulichki.ru': (6, 2),
 'geo.mozilla.org': (10, 1),
 'google.com': (1, 9),
 'mail.google.com': (5, 2),
 'meduza.io': (4, 3),
 'oracle.com': (2, 8),
 'plus.google.com': (9, 1),
 'vk.com': (3, 3),
 'yandex.ru': (11, 1)}

Примените полученную функцию к данным по 10 пользователям.

**<font color='red'> Вопрос 1. </font> Сколько уникальных сессий из 10 сайтов в выборке с 10 пользователями?**
13039

In [5]:
%%time
train_data_10users, site_freq_10users = prepare_train_set(os.path.join(PATH_TO_DATA, '10users'), 
                                                     session_length=10)

Wall time: 26.6 s


In [161]:
len(site_freq_10users)

4913

In [174]:
columns

['site1',
 'site2',
 'site3',
 'site4',
 'site5',
 'site6',
 'site7',
 'site8',
 'site9']

In [194]:
train_data_10users[columns].groupby(columns).size().shape

(13039,)

In [196]:
train_data_10users.shape[0]-train_data_10users[columns].groupby(columns).size().shape[0]

1022

**<font color='red'> Вопрос 2. </font> Сколько всего уникальных сайтов в выборке из 10 пользователей? **4913

In [197]:
''' ВАШ КОД ЗДЕСЬ '''
len(site_freq_10users)

4913

In [26]:
def prepare_train_set_2(path_to_csv_files, session_length=10):
    ''' ВАШ КОД ЗДЕСЬ '''
    dict_unique_sites = {}
    columns = list()
    for i in range(session_length):
        columns.append('site'+str(i+1))
    columns.append('user_id')
    site_index = 1
    sessions = []
    for filename in glob(path_to_csv_files+'/*.csv'):
#         print(filename)
        user_id = int(filename[-8:-4])
        data = pd.read_csv(filename)
        session_number = 0   #положение сайта в session строке
        row = [0]*10         #session строка
        row.append(user_id)
        for site in data.site:
#             print (site)
            if site not in dict_unique_sites:
                dict_unique_sites[site]=(site_index, 1)
                site_index+=1
            else:
                dict_unique_sites[site] = dict_unique_sites[site][0],dict_unique_sites[site][1]+1
            if (session_number<session_length):
                row[session_number] = dict_unique_sites[site][0]
                session_number+=1
            else:
#                 df = df.append(pd.DataFrame([row], columns=columns), ignore_index=True)
                sessions.append(row)
                row = [0]*10 + [user_id]
                session_number=0
                row[session_number] = dict_unique_sites[site][0]
                session_number+=1
        sessions.append(row)
    
    return pd.DataFrame(sessions), dict_unique_sites

In [90]:
def prepare_train_set_3(path_to_csv_files, session_length=10):
    ''' ВАШ КОД ЗДЕСЬ '''
    dict_unique_sites = {}
    columns = list()
    for i in range(session_length):
        columns.append('site'+str(i+1))
    columns.append('user_id')
    site_index = 1
    sessions = []
    for filename in glob(path_to_csv_files+'/*.csv'):
#         print(filename)
        user_id = int(filename[-8:-4])
        data = pd.read_csv(filename)
        session_number = 0   #положение сайта в session строке
        row = [0]*10         #session строка
        row.append(user_id)
        for site in data.site:
#             print (site)
            if site not in dict_unique_sites:
                dict_unique_sites[site]=(site_index, 1)
                site_index+=1
            else:
                dict_unique_sites[site] = dict_unique_sites[site][0],dict_unique_sites[site][1]+1
            if (session_number<session_length):
                row[session_number] = dict_unique_sites[site][0]
                session_number+=1
            else:
#                 df = df.append(pd.DataFrame([row], columns=columns), ignore_index=True)
                sessions.append(row)
                row = [0]*10
                row.append(user_id)
                session_number=0
                row[session_number] = dict_unique_sites[site][0]
                session_number+=1
        sessions.append(row)
    
    return pd.DataFrame(sessions), dict_unique_sites

In [75]:
%load_ext line_profiler

In [86]:
%lprun?

In [93]:
input_path = os.path.join(PATH_TO_DATA,'10users')

In [96]:
%lprun -f prepare_train_set prepare_train_set(input_path,session_length=10)

*** KeyboardInterrupt exception caught in code being profiled.

In [97]:
%lprun -f prepare_train_set_2 prepare_train_set_2(input_path,session_length=10)

In [91]:
%lprun -f prepare_train_set_3 prepare_train_set_3(input_path,session_length=10)

In [80]:
%%time
train_data_10users, site_freq_10users = prepare_train_set_2(os.path.join(PATH_TO_DATA, '10users'), 
                                                     session_length=10)

Wall time: 274 ms


In [33]:
train_data_10users.head()

0   1   2   3   4   5   6   7   8   9   10
0   1   2   3   4   3   3   4   3   5   3  31
1   6   7   8   9   3  10  11  12  13  14  31
2  14   4  14  14  15  16   6  17  18  14  31
3  19  20  19  14  14  14  14  21  22  23  31
4  24  14  15  25  26  27  28  29  30  29  31

Примените полученную функцию к данным по 150 пользователям.

**<font color='red'> Вопрос 3. </font> Сколько уникальных сессий из 10 сайтов в выборке с 150 пользователями?** 131009

In [29]:
%%time
train_data_150users, site_freq_150users = prepare_train_set_2(os.path.join(PATH_TO_DATA, '150users'), 
                                                     session_length=10)

Wall time: 3.85 s


In [34]:
train_data_150users.head()

0   1   2   3   4   5   6   7   8   9   10
0   1   1   2   3   1   2   1   3   4   5   6
1   3   6   7   4   5   5   8   9  10  10   6
2   5  11   1   1   1   1   9   5   3   1   6
3   1   1  12  12  13  12  12   1  14   1   6
4   1   1  15   1  16  17  16  17  18  15   6

In [39]:
columns = list(range(10))

In [40]:
train_data_150users[columns].groupby(columns).size().shape

(131009,)

**<font color='red'> Вопрос 4. </font> Сколько всего уникальных сайтов в выборке из 150 пользователей? ** 27797

In [42]:
''' ВАШ КОД ЗДЕСЬ '''
len(site_freq_150users)

27797

**<font color='red'> Вопрос 5. </font> Какой из этих сайтов НЕ входит в топ-10 самых популярных сайтов среди посещенных 150 пользователями?**
- www.google.fr
- www.youtube.com
- safebrowsing-cache.google.com
- <b>www.linkedin.com</b>

In [43]:
''' ВАШ КОД ЗДЕСЬ '''
site_freq_150users

{'www.google.fr': (1, 64785),
 'apis.google.com': (2, 29983),
 'www.google.com': (3, 51320),
 'cnfg.toolbarservices.com': (4, 96),
 'img.babylon.com': (5, 533),
 'ajax.googleapis.com': (6, 7811),
 'utils.delta-search.com': (7, 78),
 'go.microsoft.com': (8, 1616),
 'ieonline.microsoft.com': (9, 2639),
 'javadl-esd-secure.oracle.com': (10, 4088),
 'fpdownload2.macromedia.com': (11, 820),
 'modis.gsfc.nasa.gov': (12, 8),
 'www.nasa.gov': (13, 12),
 'clients1.google.fr': (14, 5758),
 'maps.google.fr': (15, 2580),
 'mts1.google.com': (16, 3365),
 'mts0.google.com': (17, 4068),
 'gg.google.com': (18, 2388),
 'khms0.google.fr': (19, 2133),
 'khms1.google.fr': (20, 2211),
 'mts2.google.com': (21, 790),
 'mts3.google.com': (22, 749),
 'dl.javafx.com': (23, 481),
 'accounts.google.com': (24, 13855),
 'accounts.youtube.com': (25, 4277),
 'accounts.google.fr': (26, 4053),
 'plus.google.com': (27, 18467),
 'mail.google.com': (28, 19072),
 'clients2.google.com': (29, 4785),
 'chatenabled.mail.google

In [56]:
data = sorted(site_freq_150users.items(), key=operator.itemgetter(1))

In [57]:
data

[('www.google.fr', (1, 64785)),
 ('apis.google.com', (2, 29983)),
 ('www.google.com', (3, 51320)),
 ('cnfg.toolbarservices.com', (4, 96)),
 ('img.babylon.com', (5, 533)),
 ('ajax.googleapis.com', (6, 7811)),
 ('utils.delta-search.com', (7, 78)),
 ('go.microsoft.com', (8, 1616)),
 ('ieonline.microsoft.com', (9, 2639)),
 ('javadl-esd-secure.oracle.com', (10, 4088)),
 ('fpdownload2.macromedia.com', (11, 820)),
 ('modis.gsfc.nasa.gov', (12, 8)),
 ('www.nasa.gov', (13, 12)),
 ('clients1.google.fr', (14, 5758)),
 ('maps.google.fr', (15, 2580)),
 ('mts1.google.com', (16, 3365)),
 ('mts0.google.com', (17, 4068)),
 ('gg.google.com', (18, 2388)),
 ('khms0.google.fr', (19, 2133)),
 ('khms1.google.fr', (20, 2211)),
 ('mts2.google.com', (21, 790)),
 ('mts3.google.com', (22, 749)),
 ('dl.javafx.com', (23, 481)),
 ('accounts.google.com', (24, 13855)),
 ('accounts.youtube.com', (25, 4277)),
 ('accounts.google.fr', (26, 4053)),
 ('plus.google.com', (27, 18467)),
 ('mail.google.com', (28, 19072)),
 ('cl

In [61]:
data = sorted(data, key=lambda tup: tup[1][1], reverse=True)

In [63]:
data[:10]

[('www.google.fr', (1, 64785)),
 ('www.google.com', (3, 51320)),
 ('www.facebook.com', (41, 39002)),
 ('apis.google.com', (2, 29983)),
 ('s.youtube.com', (181, 29102)),
 ('clients1.google.com', (214, 25087)),
 ('mail.google.com', (28, 19072)),
 ('plus.google.com', (27, 18467)),
 ('safebrowsing-cache.google.com', (293, 17960)),
 ('www.youtube.com', (105, 16319))]

In [66]:
site_freq_150users['vk.com']

(1508, 1323)

In [46]:
import operator
x = {1: 2, 3: 4, 4: 3, 2: 1, 0: 0}
sorted(x.items(), key=operator.itemgetter(1))

[(0, 0), (2, 1), (1, 2), (4, 3), (3, 4)]

In [47]:
import operator
x = {1: 2, 3: 4, 4: 3, 2: 1, 0: 0}
sorted(x.items(), key=operator.itemgetter(0))

[(0, 0), (1, 2), (2, 1), (3, 4), (4, 3)]

**Для дальнейшего анализа запишем полученные объекты DataFrame в csv-файлы.**

In [ ]:
train_data_10users.to_csv(os.path.join(PATH_TO_DATA, 
                                       'train_data_10users.csv'), 
                        index_label='session_id', float_format='%d')
train_data_150users.to_csv(os.path.join(PATH_TO_DATA, 
                                        'train_data_150users.csv'), 
                         index_label='session_id', float_format='%d')

## Часть 2. Работа с разреженным форматом данных

Если так подумать, то полученные признаки *site1*, ..., *site10* смысла не имеют как признаки в задаче классификации. А вот если воспользоваться идеей мешка слов из анализа текстов – это другое дело. Создадим новые матрицы, в которых строкам будут соответствовать сессии из 10 сайтов, а столбцам – индексы сайтов. На пересечении строки $i$ и столбца $j$ будет стоять число $n_{ij}$ – cколько раз сайт $j$ встретился в сессии номер $i$. Делать это будем с помощью разреженных матриц Scipy – [csr_matrix](https://docs.scipy.org/doc/scipy-0.18.1/reference/generated/scipy.sparse.csr_matrix.html). Прочитайте документацию, разберитесь, как использовать разреженные матрицы и создайте такие матрицы для наших данных. Сначала проверьте на игрушечном примере, затем примените для 10 и 150 пользователей. 

Обратите внимание, что в коротких сессиях, меньше 10 сайтов, у нас остались нули, так что первый признак (сколько раз попался 0) по смыслу отличен от остальных (сколько раз попался сайт с индексом $i$). Поэтому первый столбец разреженной матрицы надо будет удалить. 

In [18]:
X_toy, y_toy = train_data_toy.iloc[:, :-1].values, train_data_toy.iloc[:, -1].values

In [19]:
X_toy

array([[ 3,  2,  2, 10,  2,  1,  7,  5,  8,  9],
       [ 3,  1,  1,  1,  0,  0,  0,  0,  0,  0],
       [ 3,  2,  6,  6,  2,  0,  0,  0,  0,  0],
       [ 4,  1,  2,  1,  2,  1,  1,  5, 11,  4],
       [ 4,  1,  2,  0,  0,  0,  0,  0,  0,  0]])

In [ ]:
X_sparse_toy = csr_matrix ''' ВАШ КОД ЗДЕСЬ '''   

**Размерность разреженной матрицы должна получиться равной 11, поскольку в игрушечном примере 3 пользователя посетили 11 уникальных сайтов.**

In [21]:
X_sparse_toy.todense()

matrix([[1, 3, 1, 0, 1, 0, 1, 1, 1, 1, 0],
        [3, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 2, 1, 0, 0, 2, 0, 0, 0, 0, 0],
        [4, 2, 0, 2, 1, 0, 0, 0, 0, 0, 1],
        [1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0]])

In [ ]:
X_10users, y_10users = train_data_10users.iloc[:, :-1].values, \
                       train_data_10users.iloc[:, -1].values
X_150users, y_150users = train_data_150users.iloc[:, :-1].values, \
                         train_data_150users.iloc[:, -1].values

In [ ]:
X_sparse_10users = ''' ВАШ КОД ЗДЕСЬ '''
X_sparse_150users = ''' ВАШ КОД ЗДЕСЬ '''

**Сохраним эти разреженные матрицы с помощью [pickle](https://docs.python.org/2/library/pickle.html) (сериализация в Python), также сохраним вектора *y_10users, y_150users* – целевые значения (id пользователя)  в выборках из 10 и 150 пользователей. То что названия этих матриц начинаются с X и y, намекает на то, что на этих данных мы будем проверять первые модели классификации.
Наконец, сохраним также и частотные словари сайтов для 3, 10 и 150 пользователей.**

In [ ]:
with open(os.path.join(PATH_TO_DATA, 'X_sparse_10users.pkl'), 'wb') as X10_pkl:
    pickle.dump(X_sparse_10users, X10_pkl, protocol=2)
with open(os.path.join(PATH_TO_DATA, 'y_10users.pkl'), 'wb') as y10_pkl:
    pickle.dump(y_10users, y10_pkl, protocol=2)
with open(os.path.join(PATH_TO_DATA, 'X_sparse_150users.pkl'), 'wb') as X150_pkl:
    pickle.dump(X_sparse_150users, X150_pkl, protocol=2)
with open(os.path.join(PATH_TO_DATA, 'y_150users.pkl'), 'wb') as y150_pkl:
    pickle.dump(y_150users, y150_pkl, protocol=2)
with open(os.path.join(PATH_TO_DATA, 'site_freq_3users.pkl'), 'wb') as site_freq_3users_pkl:
    pickle.dump(site_freq_3users, site_freq_3users_pkl, protocol=2)
with open(os.path.join(PATH_TO_DATA, 'site_freq_10users.pkl'), 'wb') as site_freq_10users_pkl:
    pickle.dump(site_freq_10users, site_freq_10users_pkl, protocol=2)
with open(os.path.join(PATH_TO_DATA, 'site_freq_150users.pkl'), 'wb') as site_freq_150users_pkl:
    pickle.dump(site_freq_150users, site_freq_150users_pkl, protocol=2)

**Чисто для подстраховки проверим, что число столбцов в разреженных матрицах `X_sparse_10users` и `X_sparse_150users` равно ранее посчитанным числам уникальных сайтов для 10 и 150 пользователей соответственно.**

In [ ]:
assert X_sparse_10users.shape[1] == len(site_freq_10users)

In [ ]:
assert X_sparse_150users.shape[1] == len(site_freq_150users)

На следующей неделе мы еще немного поготовим данные и потестируем первые гипотезы, связанные с нашими наблюдениями. 